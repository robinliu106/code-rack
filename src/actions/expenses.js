import uuid from "uuid";
import database from "../firebase/firebase";

//ADD_SNIPPET
export const addSnippet = (expense) => ({
    type: "ADD_SNIPPET",
    expense,
});

export const startAddSnippet = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const { description = "", note = "", amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt };
        return database
            .ref(`users/${uid}/expenses`)
            .push(expense)
            .then((ref) => {
                dispatch(addSnippet({ id: ref.key, ...expense }));
            });
    };
};

//REMOVE_SNIPPET
export const removeSnippet = ({ id } = {}) => ({
    type: "REMOVE_SNIPPET",
    id,
});

export const startRemoveSnippet = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        //return async action
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeSnippet({ id }));
            });
    };
};

//EDIT_SNIPPET
export const editSnippet = (id, updates) => ({
    type: "EDIT_SNIPPET",
    id,
    updates,
});

export const startEditSnippet = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .update(updates)
            .then(() => {
                dispatch(editSnippet(id, updates));
            });
    };
};

//SET_SNIPPETS
export const setExpenses = (expenses) => ({
    type: "SET_SNIPPETS",
    expenses,
});

// export const startSetSnippets;
export const startSetSnippets = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses`)
            .once("value")
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val(),
                    });
                });
                dispatch(setExpenses(expenses));
            });
    };
};
