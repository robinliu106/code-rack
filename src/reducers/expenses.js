//Expenses Reducer
const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_SNIPPET":
            return [...state, action.expense];
        case "REMOVE_SNIPPET":
            return state.filter(({ id }) => action.id !== id);
        case "EDIT_SNIPPET":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates, //override amount (passed in through action.updates)
                    };
                } else {
                    return expense;
                }
            });
        case "SET_SNIPPETS":
            return action.expenses;

        default:
            return state;
    }
};
