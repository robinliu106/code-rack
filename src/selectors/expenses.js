import moment from "moment";

//Get visible expenses
export default (expenses, { text }) => {
    return expenses
        .filter((expense) => {
            const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
            return textMatch;
        })
        .sort((a, b) => {
            return a.createdAt < b.createdAt ? 1 : -1;
        });
};
