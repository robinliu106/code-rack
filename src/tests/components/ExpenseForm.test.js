import React from "react";
import { shallow } from "enzyme";
import ExpenseForm from "../../components/ExpenseForm";
import expenses from "../fixtures/expenses";

test("Should render ExpenseForm correctly", () => {
    const wrapper = shallow(<ExpenseForm />);
    expect(wrapper).toMatchSnapshot();
});

test("Should render ExpenseForm with expense data", () => {
    const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
    expect(wrapper).toMatchSnapshot();
});

test("Should render error for invalid form submission", () => {
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find("form").simulate("submit", {
        preventDefault: () => {},
    });
    expect(wrapper.state("error").length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test("Should set description on input change", () => {
    const wrapper = shallow(<ExpenseForm />);
    const value = "New description";
    wrapper.find("input").at(0).simulate("change", { target: { value } });
    expect(wrapper.state("description")).toBe(value);
});

test("Should set note on textarea change", () => {
    const wrapper = shallow(<ExpenseForm />);
    const value = "New note";
    wrapper.find("textarea").simulate("change", { target: { value } });
    expect(wrapper.state("note")).toBe(value);
});

test("Should set valid amount", () => {
    const wrapper = shallow(<ExpenseForm />);
    const value = "12.50";
    wrapper.find("input").at(1).simulate("change", { target: { value } });
    expect(wrapper.state("amount")).toBe(value);
});

test("Should not set invalid amount", () => {
    const wrapper = shallow(<ExpenseForm />);
    const value = "12.12.50";
    wrapper.find("input").at(1).simulate("change", { target: { value } });
    expect(wrapper.state("amount")).toBe("");
});
