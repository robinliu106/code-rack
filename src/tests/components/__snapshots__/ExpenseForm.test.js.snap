// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseForm correctly 1`] = `
"<div>
  <form onSubmit={[Function]}>
    <input type=\\"text\\" placeholder=\\"Description\\" autoFocus={true} value=\\"\\" onChange={[Function]} />
    <input type=\\"text\\" placeholder=\\"Amount\\" value=\\"\\" onChange={[Function]} />
    <SingleDatePicker date={{...}} onDateChange={[Function]} focused={false} onFocusChange={[Function]} numberOfMonths={1} isOutsideRange={[Function]} id=\\"date\\" placeholder=\\"Date\\" disabled={false} required={false} readOnly={false} screenReaderInputMessage=\\"\\" showClearDate={false} showDefaultInputIcon={false} inputIconPosition=\\"before\\" customInputIcon={{...}} customCloseIcon={{...}} orientation=\\"horizontal\\" anchorDirection=\\"left\\" openDirection=\\"down\\" horizontalMargin={0} withPortal={false} withFullScreenPortal={false} initialVisibleMonth={{...}} firstDayOfWeek={{...}} keepOpenOnDateSelect={false} reopenPickerOnClearDate={false} renderCalendarInfo={{...}} hideKeyboardShortcutsPanel={false} daySize={39} isRTL={false} navPrev={{...}} navNext={{...}} onPrevMonthClick={[Function]} onNextMonthClick={[Function]} onClose={[Function]} renderMonth={{...}} renderDay={{...}} enableOutsideDays={false} isDayBlocked={[Function]} isDayHighlighted={[Function]} displayFormat={[Function]} monthFormat=\\"MMMM YYYY\\" weekDayFormat=\\"dd\\" phrases={{...}} />
    <textarea placeholder=\\"Add a note for your expense (optional)\\" value=\\"\\" onChange={[Function]} />
    <button>
      Add Expense
    </button>
  </form>
</div>"
`;

exports[`Should render ExpenseForm with expense data 1`] = `
"<div>
  <form onSubmit={[Function]}>
    <input type=\\"text\\" placeholder=\\"Description\\" autoFocus={true} value=\\"rent\\" onChange={[Function]} />
    <input type=\\"text\\" placeholder=\\"Amount\\" value=\\"1095\\" onChange={[Function]} />
    <SingleDatePicker date={{...}} onDateChange={[Function]} focused={false} onFocusChange={[Function]} numberOfMonths={1} isOutsideRange={[Function]} id=\\"date\\" placeholder=\\"Date\\" disabled={false} required={false} readOnly={false} screenReaderInputMessage=\\"\\" showClearDate={false} showDefaultInputIcon={false} inputIconPosition=\\"before\\" customInputIcon={{...}} customCloseIcon={{...}} orientation=\\"horizontal\\" anchorDirection=\\"left\\" openDirection=\\"down\\" horizontalMargin={0} withPortal={false} withFullScreenPortal={false} initialVisibleMonth={{...}} firstDayOfWeek={{...}} keepOpenOnDateSelect={false} reopenPickerOnClearDate={false} renderCalendarInfo={{...}} hideKeyboardShortcutsPanel={false} daySize={39} isRTL={false} navPrev={{...}} navNext={{...}} onPrevMonthClick={[Function]} onNextMonthClick={[Function]} onClose={[Function]} renderMonth={{...}} renderDay={{...}} enableOutsideDays={false} isDayBlocked={[Function]} isDayHighlighted={[Function]} displayFormat={[Function]} monthFormat=\\"MMMM YYYY\\" weekDayFormat=\\"dd\\" phrases={{...}} />
    <textarea placeholder=\\"Add a note for your expense (optional)\\" value=\\"\\" onChange={[Function]} />
    <button>
      Add Expense
    </button>
  </form>
</div>"
`;

exports[`Should render error for invalid form submission 1`] = `
"<div>
  <p>
    Please provide description and amount
  </p>
  <form onSubmit={[Function]}>
    <input type=\\"text\\" placeholder=\\"Description\\" autoFocus={true} value=\\"\\" onChange={[Function]} />
    <input type=\\"text\\" placeholder=\\"Amount\\" value=\\"\\" onChange={[Function]} />
    <SingleDatePicker date={{...}} onDateChange={[Function]} focused={false} onFocusChange={[Function]} numberOfMonths={1} isOutsideRange={[Function]} id=\\"date\\" placeholder=\\"Date\\" disabled={false} required={false} readOnly={false} screenReaderInputMessage=\\"\\" showClearDate={false} showDefaultInputIcon={false} inputIconPosition=\\"before\\" customInputIcon={{...}} customCloseIcon={{...}} orientation=\\"horizontal\\" anchorDirection=\\"left\\" openDirection=\\"down\\" horizontalMargin={0} withPortal={false} withFullScreenPortal={false} initialVisibleMonth={{...}} firstDayOfWeek={{...}} keepOpenOnDateSelect={false} reopenPickerOnClearDate={false} renderCalendarInfo={{...}} hideKeyboardShortcutsPanel={false} daySize={39} isRTL={false} navPrev={{...}} navNext={{...}} onPrevMonthClick={[Function]} onNextMonthClick={[Function]} onClose={[Function]} renderMonth={{...}} renderDay={{...}} enableOutsideDays={false} isDayBlocked={[Function]} isDayHighlighted={[Function]} displayFormat={[Function]} monthFormat=\\"MMMM YYYY\\" weekDayFormat=\\"dd\\" phrases={{...}} />
    <textarea placeholder=\\"Add a note for your expense (optional)\\" value=\\"\\" onChange={[Function]} />
    <button>
      Add Expense
    </button>
  </form>
</div>"
`;
